### PR TITLE
Cleanups and Fixes for Emergency Access

### LINKED ISSUE #2925: Emergency access fails if grantor doesnt log in during waiting time
### Subject of the issue
When the grantee requests access to the grantor's vault, access isn't possible after the wait time if the grantor hasn't logged in during that time. Although the grantee received the `Emergency access request for John Doe approved` email, the `Emergency Access` tab of the grantee's account just lists the tags `Emergency Access` Initiated and `View` for the grantor user.

#### Your environment (Generated via diagnostics page)
* Vaultwarden version: v1.26.0
* Web-vault version: v2022.10.0
* Running within Docker: true (Base: Debian)
* Environment settings overridden: true
* Uses a reverse proxy: true
* IP Header check: true (X-Real-IP)
* Internet access: true
* Internet access via a proxy: false
* DNS Check: true
* Time Check: true
* Domain Configuration Check: true
* HTTPS Check: true
* Database type: SQLite
* Database version: 3.35.4
* Clients used: 
* Reverse proxy and version: 
* Other relevant information: 

### Config (Generated via diagnostics page)
<details><summary>Show Running Config</summary>

**Environment settings which are overridden:** DOMAIN, ADMIN_TOKEN, ALLOWED_IFRAME_ANCESTORS, SMTP_HOST, SMTP_SECURITY, SMTP_PORT, SMTP_FROM, SMTP_FROM_NAME, SMTP_USERNAME, SMTP_PASSWORD, SMTP_AUTH_MECHANISM

```json
{
  "_duo_akey": null,
  "_enable_duo": false,
  "_enable_email_2fa": true,
  "_enable_smtp": true,
  "_enable_yubico": true,
  "_icon_service_csp": "",
  "_icon_service_url": "",
  "_ip_header_enabled": true,
  "admin_ratelimit_max_burst": 3,
  "admin_ratelimit_seconds": 300,
  "admin_token": "***",
  "allowed_iframe_ancestors": "*****.******.**",
  "attachments_folder": "data/attachments",
  "authenticator_disable_time_drift": false,
  "data_folder": "data",
  "database_conn_init": "",
  "database_max_conns": 10,
  "database_timeout": 30,
  "database_url": "****/**.*******",
  "db_connection_retries": 15,
  "disable_2fa_remember": true,
  "disable_admin_token": false,
  "disable_icon_download": false,
  "domain": "*****://*****.******.**",
  "domain_origin": "*****://*****.******.**",
  "domain_path": "",
  "domain_set": true,
  "duo_host": null,
  "duo_ikey": null,
  "duo_skey": null,
  "email_attempts_limit": 3,
  "email_expiration_time": 600,
  "email_token_size": 6,
  "emergency_access_allowed": true,
  "emergency_notification_reminder_schedule": "0 5 * * * *",
  "emergency_request_timeout_schedule": "0 5 * * * *",
  "enable_db_wal": true,
  "extended_logging": true,
  "helo_name": null,
  "hibp_api_key": null,
  "icon_blacklist_non_global_ips": true,
  "icon_blacklist_regex": null,
  "icon_cache_folder": "data/icon_cache",
  "icon_cache_negttl": 259200,
  "icon_cache_ttl": 2592000,
  "icon_download_timeout": 10,
  "icon_redirect_code": 302,
  "icon_service": "internal",
  "incomplete_2fa_schedule": "30 * * * * *",
  "incomplete_2fa_time_limit": 3,
  "invitation_expiration_hours": 120,
  "invitation_org_name": "Bitwarden",
  "invitations_allowed": false,
  "ip_header": "X-Real-IP",
  "job_poll_interval_ms": 30000,
  "log_file": null,
  "log_level": "Info",
  "log_timestamp_format": "%Y-%m-%d %H:%M:%S.%3f",
  "login_ratelimit_max_burst": 10,
  "login_ratelimit_seconds": 60,
  "org_attachment_limit": null,
  "org_creation_users": "*****@******.**",
  "password_hints_allowed": false,
  "password_iterations": 100000,
  "reload_templates": false,
  "require_device_email": false,
  "rsa_key_filename": "data/rsa_key",
  "send_purge_schedule": "0 5 * * * *",
  "sends_allowed": true,
  "sends_folder": "data/sends",
  "show_password_hint": false,
  "signups_allowed": false,
  "signups_domains_whitelist": "******.**,*******.**",
  "signups_verify": true,
  "signups_verify_resend_limit": 6,
  "signups_verify_resend_time": 3600,
  "smtp_accept_invalid_certs": false,
  "smtp_accept_invalid_hostnames": false,
  "smtp_auth_mechanism": "\"Plain\"",
  "smtp_debug": false,
  "smtp_explicit_tls": null,
  "smtp_from": "*******@*****.******.**",
  "smtp_from_name": "Bitwarden",
  "smtp_host": "****.******.**",
  "smtp_password": "***",
  "smtp_port": 465,
  "smtp_security": "force_tls",
  "smtp_ssl": null,
  "smtp_timeout": 15,
  "smtp_username": "******************",
  "templates_folder": "data/templates",
  "tmp_folder": "data/tmp",
  "trash_auto_delete_days": null,
  "trash_purge_schedule": "0 5 0 * * *",
  "use_syslog": false,
  "user_attachment_limit": null,
  "web_vault_enabled": true,
  "web_vault_folder": "web-vault/",
  "websocket_address": "0.0.0.0",
  "websocket_enabled": true,
  "websocket_port": 3012,
  "yubico_client_id": null,
  "yubico_secret_key": null,
  "yubico_server": null
}
```

</details>

### Steps to reproduce

1. Create two users _grantor_ and _grantee_
2. _grantor_ is allowed to create organisations, _grantee_ is not
3. _grantor_ is owner in one group and _grantee_ is user the same group
4. grantor grants grantee `View` access after one day of waiting
5. grantor logs out of all of his clients
6. grantee request access from grantor
7. grantor does not log in during that waiting time
8. grantee logs in and navigates to `Emergency Access` after receiving the `Emergency access request for grantor approved` email
9. the only available option for grantee when clicking the kebab menu is `Remove`

### Expected behaviour
Grantee is able to access grantors vault after the waiting time.

### Actual behaviour
Grantee cannot access grantors vault after the waiting time.

### LINKED ISSUE #2955: Emergency access sends a lot of emails when the wait time period is reached
### Subject of the issue
Vaultwarden sends too much emails when the emergency access wait time lapses or grantor approve the request

### Deployment environment

* Vaultwarden version: v1.26.0
* Web-vault version: v2022.10.0
* Running within Docker: true (Base: Debian)
* Environment settings overridden: false
* Uses a reverse proxy: true
* IP Header check: true (X-Real-IP)
* Internet access: true
* Internet access via a proxy: false
* DNS Check: true
* Time Check: true
* Domain Configuration Check: true
* HTTPS Check: true
* Database type: SQLite
* Database version: 3.35.4

* Other relevant details:

### Steps to reproduce
Make an emergency request for another account and wait the time the grantor setted.

### Expected behaviour
Only one or some emails to inform that the emergency access is pending

### Actual behaviour
The grantor receives a lot of emails when the emergency wait time is reached

### Troubleshooting data

mails

### PR DESCRIPTION
**Pull request by BlackDex**
- Several cleanups and code optimizations for Emergency Access
- Fixed a race-condition regarding jobs for Emergency Access
- Some other small changes like `allow(clippy::)` removals

Fixes #2925
Fixes #2955

### CODE REVIEW

**Review by dani-garcia — APPROVED** [review]

File: .env.template [hunk 1]
```diff
@@ -102,12 +102,12 @@
 # INCOMPLETE_2FA_SCHEDULE="30 * * * * *"
 ##
 ## Cron schedule of the job that sends expiration reminders to emergency access grantors.
-## Defaults to hourly (5 minutes after the hour). Set blank to disable this job.
-# EMERGENCY_NOTIFICATION_REMINDER_SCHEDULE="0 5 * * * *"
+## Defaults to hourly (3 minutes after the hour). Set blank to disable this job.
+# EMERGENCY_NOTIFICATION_REMINDER_SCHEDULE="0 3 * * * *"
```
Comment (tessus) [inline .env.template:106]:
```suggestion
# EMERGENCY_NOTIFICATION_REMINDER_SCHEDULE="3 * * * *"
```

If this is a real crontab entry, then the current format is wrong. Crontab has 5 time/date fields and one field for the command.

Every hour at minute 3 is `3 * * * * command`

File: .env.template [hunk 2]
```diff
@@ -102,12 +102,12 @@
...
 ##
 ## Cron schedule of the job that grants emergency access requests that have met the required wait time.
-## Defaults to hourly (5 minutes after the hour). Set blank to disable this job.
-# EMERGENCY_REQUEST_TIMEOUT_SCHEDULE="0 5 * * * *"
+## Defaults to hourly (7 minutes after the hour). Set blank to disable this job.
+# EMERGENCY_REQUEST_TIMEOUT_SCHEDULE="0 7 * * * *"
```
Comment (tessus) [inline .env.template:110]:
```suggestion
# EMERGENCY_REQUEST_TIMEOUT_SCHEDULE="7 * * * *"
```

If this is a real crontab entry, then the current format is wrong. Crontab has 5 time/date fields and one field for the command.

Every hour at minute 7 is `7 * * * * command`

**Review by tessus — COMMENTED** [review]
I checked the other cron jobs as well and all of them are wrong.


Cron format:

```
*    *    *    *    *  Command_to_execute
|    |    |    |    |       
|    |    |    |    Day of the Week ( 0 - 6 ) ( Sunday = 0 )
|    |    |    |
|    |    |    Month ( 1 - 12 )
|    |    |
|    |    Day of Month ( 1 - 31 )
|    |
|    Hour ( 0 - 23 )
|
Min ( 0 - 59 )
```

According to @stefan0xC, the format is different than the standard cron format.

I'll open a PR to add this info to the template file. This is certainly not clear by just looking at the file....

File: .env.template (hunk 2 above)
Comment (stefan0xC) [reply .env.template:110]: The [cron](https://github.com/zslayton/cron) library actually uses 7 fields (but the year might be omitted):
```
sec   min   hour   day of month   month   day of week   year
*     *     *      *              *       *             *
````

C.f. the documentation of the crate Vaultwarden uses: https://docs.rs/job_scheduler_ng/2.0.2/job_scheduler_ng/

Comment (tessus) [reply .env.template:110]: Ok, this is not a real cron format. I'll delete my comments then.

**Comment by BlackDex** [comment]
> Update: Scratch that too. It is in the `.env.template` file. Strange that I missed this.

That was in one of my updates, so it could be you didn't see a recent version.
Because of the difference between normal cron's and sometimes the complexity i thought to add some more info about it.
The seconds option is btw a nice addition, at least for me during testing, instead of having to wait a whole minute or 30 seconds hehe.

**Comment by tessus** [comment]
> That was in one of my updates, so it could be you didn't see a recent version.

Yep, this is possible, because I remember I read the entire template file a while back and that part was new to me.

> Because of the difference between normal cron's and sometimes the complexity i thought to add some more info about it.

Good idea. It definitely helps.

> The seconds option is btw a nice addition, at least for me during testing, 

Yep, I totally agree. Personally I would have just added new fields at the end, so that it is still compatible with the original format. But I also understand why the dev of the library wanted to have the seconds next to the minutes...

**Pull request merged**